From 8ea588be50b59c5b31291906ed42b124c9c7323d Mon Sep 17 00:00:00 2001
From: Alexandru Ardelean <alexandru.ardelean@analog.com>
Date: Wed, 17 Jul 2019 17:30:17 +0300
Subject: [PATCH 0650/1851] uio: rename uio.c -> uio_core.c & unify into uio.ko

commit b62591048a35e571cb5405d13c4704771fdb0947 from
https://github.com/Xilinx/linux-xlnx.git

When uio_dmabuf.c was added, it broke building uio as a kmod.

The error is:
```
ERROR: "uio_dmabuf_unmap" [drivers/uio/uio.ko] undefined!
ERROR: "uio_dmabuf_map" [drivers/uio/uio.ko] undefined!
ERROR: "uio_dmabuf_cleanup" [drivers/uio/uio.ko] undefined!
```

Two kmods would be built: `uio.ko` & `uio_dmabuf.ko`.
The nasty part is that `uio.ko` would require symbols from `uio_dmabuf.ko`,
which isn't ideal since `uio.c` is the core part of the UIO subsystem.

Exporting the above symbols is not ideal either, because even if it fixes
the build, it fails at runtime with:
```
[    7.594314] uio_dmabuf: Unknown symbol dma_buf_get (err -2)
[    7.594344] uio_dmabuf: Unknown symbol dma_buf_put (err -2)
[    7.594412] uio_dmabuf: Unknown symbol dma_buf_detach (err -2)
[    7.639590] uio_dmabuf: Unknown symbol dma_buf_attach (err -2)
[    7.639636] uio_dmabuf: Unknown symbol dma_buf_map_attachment (err -2)
[    7.639673] uio_dmabuf: Unknown symbol dma_buf_unmap_attachment (err -2)
[    7.639712] uio_dmabuf: Unknown symbol dma_buf_get (err -2)
[    7.639742] uio_dmabuf: Unknown symbol dma_buf_put (err -2)
```

The simplest (and not necessarily the most ideal) fix is to rename `uio.c`
to `uio_core.c`, and make sure that `uio.ko` includes both sources into
that single kmod.
The alternative would have been to move all the code in the `uio.c` file.

And the other alternative would be to make sure that that `uio_dmabuf.ko`
can be more self-sufficient.

Fixes: de2151831d9 ("uio: Add dma-buf import ioctls")
Signed-off-by: Alexandru Ardelean <alexandru.ardelean@analog.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/uio/Makefile              | 4 +++-
 drivers/uio/{uio.c => uio_core.c} | 0
 2 files changed, 3 insertions(+), 1 deletion(-)
 rename drivers/uio/{uio.c => uio_core.c} (100%)

diff --git a/drivers/uio/Makefile b/drivers/uio/Makefile
index 14636e0c8ef0..fe354273faf9 100644
--- a/drivers/uio/Makefile
+++ b/drivers/uio/Makefile
@@ -1,5 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
-obj-$(CONFIG_UIO)	+= uio.o uio_dmabuf.o
+uio-y := uio_core.o uio_dmabuf.o
+
+obj-$(CONFIG_UIO)	+= uio.o
 obj-$(CONFIG_UIO_CIF)	+= uio_cif.o
 obj-$(CONFIG_UIO_PDRV_GENIRQ)	+= uio_pdrv_genirq.o
 obj-$(CONFIG_UIO_DMEM_GENIRQ)	+= uio_dmem_genirq.o
diff --git a/drivers/uio/uio.c b/drivers/uio/uio_core.c
similarity index 100%
rename from drivers/uio/uio.c
rename to drivers/uio/uio_core.c
-- 
2.31.1

